packages/gatsby/src/query/query-compiler.js

```diff
@@ -1,384 +1,400 @@
+}
 
-    for (let [filePath, doc] of nodes.entries()) {
-      let errors = validate(this.schema, doc, validationRules)
+export const processQueries = ({
+  schema,
+  parsedQueries,
+  addError,
+  parentSpan,
+}) => {
+  const { definitionsByName, operations } = extractOperations(
+    schema,
+    parsedQueries,
+    addError,
+    parentSpan
+  )
 
-      if (errors && errors.length) {
-        addError(
-          ...errors.map(error => {
-            const location = {
-              start: locInGraphQlToLocInFile(
-                doc.definitions[0].templateLoc,
-                error.locations[0]
-              ),
-            }
-            return errorParser({ message: error.message, filePath, location })
-          })
-        )
+  return processDefinitions({
+    schema,
+    operations,
+    definitionsByName,
+    addError,
+    parentSpan,
+  })
+}
 
-        boundActionCreators.queryExtractionGraphQLError({
-          componentPath: filePath,
-        })
-        return compiledNodes
-      }
+const preValidationRules = [
+  LoneAnonymousOperationRule,
+  KnownTypeNamesRule,
+  FragmentsOnCompositeTypesRule,
+  VariablesAreInputTypesRule,
+  ScalarLeafsRule,
+  PossibleFragmentSpreadsRule,
+  ValuesOfCorrectTypeRule,
+  VariablesInAllowedPositionRule,
+]
 
-      // The way we currently export fragments requires duplicated ones
-      // to be filtered out since there is a global Fragment namespace
-      // We maintain a top level fragment Map to keep track of all definitions
-      // of the fragment type and to filter them out if theythey've already been
-      // declared before
-      doc.definitions = doc.definitions.filter(definition => {
-        if (definition.kind === Kind.FRAGMENT_DEFINITION) {
-          const fragmentName = definition.name.value
-          if (fragmentMap.has(fragmentName)) {
-            if (print(definition) === fragmentMap.get(fragmentName)) {
-              return false
-            }
-          } else {
-            fragmentMap.set(fragmentName, print(definition))
+const extractOperations = (schema, parsedQueries, addError, parentSpan) => {
+  const definitionsByName = new Map()
+  const operations = []
+
+  for (const [filePath, doc] of parsedQueries.entries()) {
+    const errors = validate(schema, doc, preValidationRules)
+
+    if (errors && errors.length) {
+      addError(
+        ...errors.map(error => {
+          const location = {
+            start: locInGraphQlToLocInFile(
+              doc.definitions[0].templateLoc,
+              error.locations[0]
+            ),
           }
-        }
-        return true
-      })
+          return errorParser({ message: error.message, filePath, location })
+        })
+      )
 
-      documents.push(doc)
-      doc.definitions.forEach((def: any) => {
-        const name: string = def.name.value
-        namePathMap.set(name, filePath)
-        nameDefMap.set(name, def)
+      actions.queryExtractionGraphQLError({
```

Comment:
I'm not sure but isn't `actions.queryExtractionGraphQLError` just return an object that you still need to dispatch yourself?

https://github.com/gatsbyjs/gatsby/blob/7271a8d7cf2ee26370e3fa9474b8146af47eebb6/packages/gatsby/src/redux/actions/internal.js#L117-L128